src/Neo.VM/ReferenceCounterV2.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (C) 2015-2024 The Neo Project.
+//
+// ReferenceCounterV2.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.VM.Types;
+using System;
+
+namespace Neo.VM
+{
+    /// <summary>
+    /// Used for reference counting of objects in the VM.
+    /// </summary>
+    public sealed class ReferenceCounterV2 : IReferenceCounter
+    {
+        /// <summary>
+        /// Gets the count of references.
+        /// </summary>
+        public int Count { get; private set; }
+
+        /// <summary>
+        /// Adds item to Reference Counter
+        /// </summary>
+        /// <param name="item">The item to add.</param>
+        /// <param name="count">Number of similar entries</param>
+        public void AddStackReference(StackItem item, int count = 1)
+        {
+            Count += count;
+        }
+
+        /// <summary>
+        /// Removes item from Reference Counter
+        /// </summary>
+        /// <param name="item">The item to remove.</param>
+        public void RemoveStackReference(StackItem item)
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Reference was not added before");
+            }
+
+            Count--;
```

Comment:
Hi, what if you are removing a compound type? for instance, you have  an array with 2047 sub items, then you remove the array, the rc will only remove one reference, and the vm will still be full.

Reply:
```C#
        [TestMethod]
        public void TestNewArrayThenDrop()
        {
            using ScriptBuilder sb = new();
            sb.EmitPush(2040);
            sb.Emit(OpCode.NEWARRAY);
            sb.Emit(OpCode.DROP);
            using ExecutionEngine engine = new();
            engine.LoadScript(sb.ToArray());
            Assert.AreEqual(0, engine.ReferenceCounter.Count);
            Assert.AreEqual(VMState.HALT, engine.Execute());
            Assert.AreEqual(0, engine.ReferenceCounter.Count);
        }
```

Assert.AreEqual failed. Expected:<0>. Actual:<2040>. 
   at Neo.Test.UT_ReferenceCounter.TestNewArrayThenDrop() in /Users/jinghuiliao/git/neo/tests/Neo.VM.Tests/UT_ReferenceCounter.cs:line 257
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)




op:[0000]PUSHINT16 []
op:[0003]NEWARRAY [Integer(2040)]
op:[0004]DROP [Array(2040)]
op:[0005] []

Reply:
```C#
        [TestMethod]
        public void TestNewArrayThenDrop()
        {
            using ScriptBuilder sb = new();
            sb.EmitPush(2040);
            sb.Emit(OpCode.NEWARRAY);
            sb.Emit(OpCode.DROP);
            sb.EmitPush(10);
            sb.Emit(OpCode.NEWARRAY);
            using ExecutionEngine engine = new();
            engine.LoadScript(sb.ToArray());
            Assert.AreEqual(0, engine.ReferenceCounter.Count);
            var res = engine.Execute();
            Assert.AreEqual(0, engine.ReferenceCounter.Count);
            Assert.AreEqual(VMState.HALT, res);
        }
```

Assert.AreEqual failed. Expected:<0>. Actual:<2051>. 
   at Neo.Test.UT_ReferenceCounter.TestNewArrayThenDrop() in /Users/jinghuiliao/git/neo/tests/Neo.VM.Tests/UT_ReferenceCounter.cs:line 259
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)




op:[0000]PUSHINT16 []
op:[0003]NEWARRAY [Integer(2040)]
op:[0004]DROP [Array(2040)]
op:[0005]PUSH10 []
op:[0006]NEWARRAY [Integer(10)]

Reply:
> Hi, what if you are removing a compound type? for instance, you have an array with 2047 sub items, then you remove the array, the rc will only remove one reference, and the vm will still be full.

yes, we need to propagate this change

Reply:
Could you add these UT?

Reply:
```csharp
  Assert.AreEqual(0, engine.ReferenceCounter.Count);
            Assert.AreEqual(VMState.HALT, res);
```
            it's 11 in the previous version

Reply:
> ```cs
>         [TestMethod]
>         public void TestNewArrayThenDrop()
>         {
>             using ScriptBuilder sb = new();
>             sb.EmitPush(2040);
>             sb.Emit(OpCode.NEWARRAY);
>             sb.Emit(OpCode.DROP);
>             using ExecutionEngine engine = new();
>             engine.LoadScript(sb.ToArray());
>             Assert.AreEqual(0, engine.ReferenceCounter.Count);
>             Assert.AreEqual(VMState.HALT, engine.Execute());
>             Assert.AreEqual(0, engine.ReferenceCounter.Count);
>         }
> ```
> 
> Assert.AreEqual failed. Expected:<0>. Actual:<2040>. at Neo.Test.UT_ReferenceCounter.TestNewArrayThenDrop() in /Users/jinghuiliao/git/neo/tests/Neo.VM.Tests/UT_ReferenceCounter.cs:line 257 at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor) at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
> 
> op:[0000]PUSHINT16 [] op:[0003]NEWARRAY [Integer(2040)] op:[0004]DROP [Array(2040)] op:[0005] []

this use the previous version also

Reply:
> > Hi, what if you are removing a compound type? for instance, you have an array with 2047 sub items, then you remove the array, the rc will only remove one reference, and the vm will still be full.
> 
> yes, we need to propagate this change

Its not about propagate, its that this will make it almost impossible to run contracts, any contract that ever create over 2048 items either in one compoundtype or multiple compoundtype will be not able to work anymore. This is like we allow at most 2048 items ever being created.